[FSDP][7/N] Support `replicate` in `fully_shard`

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#91044 [FSDP][7/N] Support `replicate` in `fully_shard`**
* #90959 [FSDP][6/N] Add note explaining idioms for `_FSDPState` traversal
* #90874 [FSDP][5/N] Add manual "wrapping" support for `fully_shard`
* #90871 [FSDP][4/N] Refactor func to share state/init handle attrs
* #90945 fully_shard load state_dict

This PR supports nesting `replicate` in `fully_shard`.
- The PR achieves this by treating `replicate`-annotated modules are ignored modules. This means that all submodules in the `replicate`-annotated module's subtree are ignored, including nested `fully_shard`-annotated modules, which is the desired behavior.

---

This PR reworks some tree traversal.

One end goal is for `state._handles` to follow the same order for both the wrapper and composable paths. This implies that `_get_fsdp_handles()` returns the same value for both paths.
- The helper function `_get_fully_sharded_module_to_states()` now follows a left-to-right DFS from each fully sharded module instead of a BFS. The left-to-right DFS follows `.modules()` order.
- The composable auto "wrap" initialization function `_init_param_handles_from_module()` follows the reverse left-to-right DFS order. As noted in the code comments, this initialization order is a valid reverse topological sort, but it differs from the wrapper path. This is the _only_ difference with respect to initialization order through the entire process.
```
mod: Module(
    submod1: Submodule()
    submod2: Submodule(
        subsubmod: Subsubmodule(),
    ),
)
```
For left-to-right DFS, the order is `mod`, `submod1`, `submod2`, `subsubmod`. (For context, right-to-left DFS would be `mod`, `submod2`, `subsubmod`, `submod1`. In other words, the left-to-right vs. right-to-left corresponds to `.children()` vs. `reversed(.children())` respectively.) Then, reverse left-to-right DFS is `subsubmod`, `submod2`, `submod1`, `mod`, which is a valid initialization order. However, the wrapper auto wrap initialization order would be `submod1`, `subsubmod`, `submod2`, `mod` since it directly follows a left-to-right DFS and initializes as a part of the recursive DFS logic.
- At the end of `_init_param_handles_from_module()`, we reverse the newly populated `state._handles`, so this is the reverse reverse left-to-right DFS order, which is equivalent to the left-to-right DFS order. Thus, `state._handles` has the same order for both paths.

Another goal is for `_get_fsdp_states()` to not traverse into any submodule that is annotated with an API that is not compatible with `fully_shard` (e.g. `replicate`). To achieve this while preserving that `_get_fsdp_states()` follows `.modules()` order, we again use a left-to-right DFS.

The reason the DFSs may look strange is because I implemented them non-recursively, which requires a stack.

- `test_get_fully_sharded_module_to_states()` in `test_utils.py` checks the traversal order of `_get_fully_sharded_module_to_states()`.
- `test_policy()` in `test_fully_shard.py` checks the traversal order returned by `_get_fsdp_handles()`.

---

Due to a circular dependency issue, we must move the graph/tree traversal helpers to their own file `_traversal_utils.py`, and any usages must import the entire file like `import torch.distributed.fsdp._traversal_utils as traversal_utils` instead of `from torch.distributed.fsdp._traversal_utils import ...`.

The cycle comes from the fact that the traversals require `_composable()`, which requires `_get_registry()` from `composable/contract.py`, which when imported, imports `composable/fully_shard.py`, which requires the traversals.

Commit:
[FSDP][7/N] Support `replicate` in `fully_shard`

[ghstack-poisoned]

Commit:
Update on "[FSDP][7/N] Support `replicate` in `fully_shard`"




[ghstack-poisoned]

Commit:
Update on "[FSDP][7/N] Support `replicate` in `fully_shard`"




[ghstack-poisoned]

Commit:
Update on "[FSDP][7/N] Support `replicate` in `fully_shard`"


This PR needs to be rebased to get the newly landed PRs.

[ghstack-poisoned]

Commit:
Update on "[FSDP][7/N] Support `replicate` in `fully_shard`"


This PR needs to be rebased to get the newly landed PRs.

I will update with a proper PR summary before requesting for review.

[ghstack-poisoned]

Commit:
Update on "[FSDP][7/N] Support `replicate` in `fully_shard`"


This PR needs to be rebased to get the newly landed PRs.

I will update with a proper PR summary before requesting for review.

[ghstack-poisoned]

Commit:
Update on "[FSDP][7/N] Support `replicate` in `fully_shard`"


This PR needs to be rebased to get the newly landed PRs.

I will update with a proper PR summary before requesting for review.

[ghstack-poisoned]

mrshenli commented on torch/distributed/fsdp/_composable_utils.py:
```diff
@@ -0,0 +1,94 @@
+"""
+NOTE: This file must be imported like
```
trying to make sure I understand the circular dependency here. Basically `torch.distributed.fsdp` files imported features from `_composable`, and `_composable.fully_shard` imported `torch.distributed.fsdp` utils, which created a circle. We can address this by moving the utils like `_get_registry` to a 3rd package, but I am not sure yet if that makes sense. Let's use your current solution in this PR. We can come back to this later. We basically need a clear dependency direction between composable features and wrapper features.

awgu replied:
When I was reading the error tracebacks, I saw that importing from `_composable/contract.py` imports `_composable/fully_shard.py` since `_composable/__init__.py` imports `fully_shard.py`. Therefore, the cycle is as you described: Some utils in `fsdp/` need to import from `_composable/contract.py` (for `_get_registry`), while other utils in `fsdp/` are imported from `_composable/fully_shard.py` (as expected).

The current solution is to isolate the utils in `fsdp/` that need to import from `_composable/contract.py` for `_get_registry`. This is okay for now since it is only three functions that are moved. They now exist in `fsdp/_traversal_utils.py`.  The downside is using any of those functions in another file in `fsdp/` requires something like `import torch.distributed.fsdp._traversal_utils as traversal_utils` and `traversal_utils._get_fsdp_states` each time.

mrshenli commented on torch/distributed/fsdp/_composable_utils.py:
```diff
@@ -0,0 +1,94 @@
+"""
```
curious about the naming of this file. This seems to be used by both composable and wrapper features. Any reason we only name this file with a composable prefix?

awgu replied:
Good point! This name does not make sense. Let me rename it to `traversal_utils`.

mrshenli commented on torch/distributed/fsdp/_init_utils.py:
```diff
@@ -533,7 +539,7 @@ def _get_ignored_modules(
+    ignored_modules.update(
         child
         for module in ignored_root_modules
         for child in module.modules()
```
Just to make sure this is expected:  if `_ignored_modules ` contains a `fully_shard` child, the subtree rooted at `fully_shard` will be added to `ignored_modules`. However, if `_ignored_modules` is empty or None, only the `fully_shard` module will be added?

awgu replied:
> if `_ignored_modules`  contains a `fully_shard` child, the subtree rooted at `fully_shard` will be added to `ignored_modules`.

Yes.

> However, if `_ignored_modules` is empty or `None`, only the `fully_shard` module will be added?

I am not sure why the `fully_shard` module would be added to `ignored_modules`? Should it?

---

I also reworked the logic to ignore modules since your comment (though I do not think it affected your questions). The semantics now are:
- Passing `ignored_modules=None` is equivalent to passing `ignored_modules=[]`.
- FSDP converts the ignored modules into a set to de-dup.
- FSDP adds any modules that cannot compose with `fully_shard` to the set. Note that in the wrapper path, we do not expect any `replicate` annotations, so this step should be a no-op.
- The remaining logic is as before. We recurse over all module subtrees for the modules in the ignored modules set (including those annotated with `replicate` for example) and ignore the entire subtree. One caveat is that `FullyShardedDataParallel` modules are not included; this is from existing logic. I think we should still include nested `fully_shard`-annotated modules if they exist in an ignored module's subtree. This is how we want to compose `fully_shard` above `replicate` above `fully_shard`.
- We also include any ignored modules already computed by child `fully_shard` or `FullyShardedDataParallel` modules.

The last two steps seem to have redundancy, but we would need to explore that more in a follow-up.

mrshenli commented on torch/distributed/fsdp/_wrap_utils.py:
```diff
@@ -129,16 +129,24 @@ def _get_submodule_to_states(
-        # not already associated with another module in `wrapped_modules`.
-        queue: Deque[Tuple[nn.Module, str]] = collections.deque()
-        queue.append((submodule, ""))
+        # Perform a DFS from `submodule` and record all unvisited state that is
```
I assume state dict save/load does not care about this, so changing the order here won't break BC?

awgu replied:
There are some subtle things happening with the traversal orders throughout this PR, which led to me having to rework some logic.

For this particular question, this should not affect state dict save and load. The only place where traversal order matters is which traversal order to use to have `state._handles` be ordered the same across wrapper and composable paths (which does not affect correctness but can improve the debugging experience). I had to rework some logic to achieve this. I will explain more details in the PR summary.

mrshenli reviewed (approved):
Thanks for making super fast progress on this!

Commit:
Update on "[FSDP][7/N] Support `replicate` in `fully_shard`"


This PR needs to be rebased to get the newly landed PRs.

I will update with a proper PR summary before requesting for review.

[ghstack-poisoned]

awgu commented on test/distributed/fsdp/test_utils.py:
```diff
@@ -169,11 +169,11 @@ def test_module_wrap_policy(self):
         """
         model = self.Model(TestGetSubmoduleToStates.SharedParameterMode.PARENT_CHILD)
 
```
These changes are just renaming from `submodule` to `fully_sharded_module` (missed from the previous PR introducing "fully sharded module").

Commit:
Update on "[FSDP][7/N] Support `replicate` in `fully_shard`"


This PR needs to be rebased to get the newly landed PRs.

I will update with a proper PR summary before requesting for review.

[ghstack-poisoned]

Commit:
Update on "[FSDP][7/N] Support `replicate` in `fully_shard`"


This PR needs to be rebased to get the newly landed PRs.

I will update with a proper PR summary before requesting for review.

[ghstack-poisoned]

Commit:
Update on "[FSDP][7/N] Support `replicate` in `fully_shard`"


This PR needs to be rebased to get the newly landed PRs.

I will update with a proper PR summary before requesting for review.

[ghstack-poisoned]

Commit:
Update on "[FSDP][7/N] Support `replicate` in `fully_shard`"


This PR needs to be rebased to get the newly landed PRs.

I will update with a proper PR summary before requesting for review.

[ghstack-poisoned]

Commit:
Update on "[FSDP][7/N] Support `replicate` in `fully_shard`"


This PR needs to be rebased to get the newly landed PRs.

I will update with a proper PR summary before requesting for review.

[ghstack-poisoned]

awgu commented on test/distributed/fsdp/test_fsdp_overlap.py:
```diff
@@ -93,9 +93,9 @@ def world_size(self):
     def _dist_train(self):
         rank = self.rank
         world_size = self.world_size
```
This test is disabled in CI, but I found that it is broken when running locally. This change fixes the test.

Commit:
Update on "[FSDP][7/N] Support `replicate` in `fully_shard`"


This PR needs to be rebased to get the newly landed PRs.

I will update with a proper PR summary before requesting for review.

[ghstack-poisoned]

Commit:
Update on "[FSDP][7/N] Support `replicate` in `fully_shard`"


This PR needs to be rebased to get the newly landed PRs.

I will update with a proper PR summary before requesting for review.

[ghstack-poisoned]

Commit:
Update on "[FSDP][7/N] Support `replicate` in `fully_shard`"


This PR supports nesting `replicate` in `fully_shard`.
- The PR achieves this by treating `replicate`-annotated modules are ignored modules. This means that all submodules in the `replicate`-annotated module's subtree are ignored, including nested `fully_shard`-annotated modules, which is the desired behavior.

---

This PR reworks some tree traversal.

One end goal is for `state._handles` to follow the same order for both the wrapper and composable paths. This implies that `_get_fsdp_handles()` returns the same value for both paths.
- The helper function `_get_fully_sharded_module_to_states()` now follows a left-to-right DFS from each fully sharded module instead of a BFS. The left-to-right DFS follows `.modules()` order.
- The composable auto "wrap" initialization function `_init_param_handles_from_module()` follows the reverse left-to-right DFS order. As noted in the code comments, this initialization order is a valid reverse topological sort, but it differs from the wrapper path. This is the _only_ difference with respect to initialization order through the entire process.
```
mod: Module(
    submod1: Submodule()
    submod2: Submodule(
        subsubmod: Subsubmodule(),
    ),
)
```
For left-to-right DFS, the order is `mod`, `submod1`, `submod2`, `subsubmod`. (For context, right-to-left DFS would be `mod`, `submod2`, `subsubmod`, `submod1`. In other words, the left-to-right vs. right-to-left corresponds to `.children()` vs. `reversed(.children())` respectively.) Then, reverse left-to-right DFS is `subsubmod`, `submod2`, `submod1`, `mod`, which is a valid initialization order. However, the wrapper auto wrap initialization order would be `submod1`, `subsubmod`, `submod2`, `mod` since it directly follows a left-to-right DFS and initializes as a part of the recursive DFS logic.
- At the end of `_init_param_handles_from_module()`, we reverse the newly populated `state._handles`, so this is the reverse reverse left-to-right DFS order, which is equivalent to the left-to-right DFS order. Thus, `state._handles` has the same order for both paths.

Another goal is for `_get_fsdp_states()` to not traverse into any submodule that is annotated with an API that is not compatible with `fully_shard` (e.g. `replicate`). To achieve this while preserving that `_get_fsdp_states()` follows `.modules()` order, we again use a left-to-right DFS.

The reason the DFSs may look strange is because I implemented them non-recursively, which requires a stack.

- `test_get_fully_sharded_module_to_states()` in `test_utils.py` checks the traversal order of `_get_fully_sharded_module_to_states()`.
- `test_policy()` in `test_fully_shard.py` checks the traversal order returned by `_get_fsdp_handles()`.

---

Due to a circular dependency issue, we must move the graph/tree traversal helpers to their own file `_traversal_utils.py`, and any usages must import the entire file like `import torch.distributed.fsdp._traversal_utils as traversal_utils` instead of `from torch.distributed.fsdp._traversal_utils import ...`.

The cycle comes from the fact that the traversals require `_composable()`, which requires `_get_registry()` from `composable/contract.py`, which when imported, imports `composable/fully_shard.py`, which requires the traversals.

[ghstack-poisoned]

Commit:
Update on "[FSDP][7/N] Support `replicate` in `fully_shard`"


This PR supports nesting `replicate` in `fully_shard`.
- The PR achieves this by treating `replicate`-annotated modules are ignored modules. This means that all submodules in the `replicate`-annotated module's subtree are ignored, including nested `fully_shard`-annotated modules, which is the desired behavior.

---

This PR reworks some tree traversal.

One end goal is for `state._handles` to follow the same order for both the wrapper and composable paths. This implies that `_get_fsdp_handles()` returns the same value for both paths.
- The helper function `_get_fully_sharded_module_to_states()` now follows a left-to-right DFS from each fully sharded module instead of a BFS. The left-to-right DFS follows `.modules()` order.
- The composable auto "wrap" initialization function `_init_param_handles_from_module()` follows the reverse left-to-right DFS order. As noted in the code comments, this initialization order is a valid reverse topological sort, but it differs from the wrapper path. This is the _only_ difference with respect to initialization order through the entire process.
```
mod: Module(
    submod1: Submodule()
    submod2: Submodule(
        subsubmod: Subsubmodule(),
    ),
)
```
For left-to-right DFS, the order is `mod`, `submod1`, `submod2`, `subsubmod`. (For context, right-to-left DFS would be `mod`, `submod2`, `subsubmod`, `submod1`. In other words, the left-to-right vs. right-to-left corresponds to `.children()` vs. `reversed(.children())` respectively.) Then, reverse left-to-right DFS is `subsubmod`, `submod2`, `submod1`, `mod`, which is a valid initialization order. However, the wrapper auto wrap initialization order would be `submod1`, `subsubmod`, `submod2`, `mod` since it directly follows a left-to-right DFS and initializes as a part of the recursive DFS logic.
- At the end of `_init_param_handles_from_module()`, we reverse the newly populated `state._handles`, so this is the reverse reverse left-to-right DFS order, which is equivalent to the left-to-right DFS order. Thus, `state._handles` has the same order for both paths.

Another goal is for `_get_fsdp_states()` to not traverse into any submodule that is annotated with an API that is not compatible with `fully_shard` (e.g. `replicate`). To achieve this while preserving that `_get_fsdp_states()` follows `.modules()` order, we again use a left-to-right DFS.

The reason the DFSs may look strange is because I implemented them non-recursively, which requires a stack.

- `test_get_fully_sharded_module_to_states()` in `test_utils.py` checks the traversal order of `_get_fully_sharded_module_to_states()`.
- `test_policy()` in `test_fully_shard.py` checks the traversal order returned by `_get_fsdp_handles()`.

---

Due to a circular dependency issue, we must move the graph/tree traversal helpers to their own file `_traversal_utils.py`, and any usages must import the entire file like `import torch.distributed.fsdp._traversal_utils as traversal_utils` instead of `from torch.distributed.fsdp._traversal_utils import ...`.

The cycle comes from the fact that the traversals require `_composable()`, which requires `_get_registry()` from `composable/contract.py`, which when imported, imports `composable/fully_shard.py`, which requires the traversals.

[ghstack-poisoned]

awgu commented:
@pytorchbot merge
